File: hstt/scripts/camerad-perf/py/TimingJitter_runner.py
#####
# Basic tests of the ZMQ-based camerad readout timing
#
# This file is based on xsub-socket.py by Mike
#####

import zmq
import time
import numpy as np
import time
import matplotlib.pyplot as plt

NS2US = 1e-3        # nanosecond to microsecond conversion
NS2SEC = 1e-9       # nanosecond to second conversion

# User Inputs
Nsamp = 5000       # Size of preallocated timestamps array (max number of samples)
isPrint = False     # Flag to print receive messages to terminal
isPlot = True       # Flag to plot the results

def main():
    # Create a ZeroMQ context
    context = zmq.Context()

    # Create a PULL socket
    socket = context.socket(zmq.XSUB)

    # Bind the socket to port 5555
    socket.connect("tcp://localhost:5555")

    # Subscribe to all topics
    socket.send(b'\x01')
    print("Listening for messages on port 5555...")

    # Preallocate array to hold msg read timestamps
    timestamps = np.empty(Nsamp, dtype=int)
    ctr = 0

    # Continuously listen for incoming messages
    try:
        while True:
            try:
                message = socket.recv()
                timestamps[ctr] = time.perf_counter_ns()
                if isPrint: 
                    print(f'Time: {timestamps[ctr]} - Raw data:  {message}')

            except zmq.Again:
                timestamps[ctr] = -1*time.perf_counter_ns()
                time.sleep(5)
                # print("Waiting for messages...")
            
            ctr += 1
            # Bailout if the next read will exceed user-requested samples
            if ctr >= Nsamp:
                ctr -= 1
                break
    except KeyboardInterrupt as e:
        print('Ctrl-C detected --- exiting reader loop')

    return timestamps[:ctr]

if __name__ == "__main__":
    timestamps = main()
    timedeltas = np.diff(timestamps) * NS2US

    # Compute statistics
    STD     = np.std(timedeltas)
    Mean    = np.mean(timedeltas)
    Median   = np.median(timedeltas)

    ######## Printed statistics
    if any(timestamps < 0):
        Nneg = np.count_nonzero(timestamps<0)
        print(f"\nWARNING:::\n\t{Nneg} timestamps were negative, meaning recv failed sometimes")

    print(f'\nSamples: {len(timestamps)} | STD: {STD:0.3f} us | Mean: {Mean:0.3f} us | Media: {Median:0.3f}')

    # Check if any timedeltas were far out of form (major outliers)
    outSig = 8      # tolerance for outliers (number of +sigma from mean)
    outliers = np.where(timedeltas > Mean + outSig*STD)[0]
    if len(outliers):
        print(f"{len(outliers)} outliers (+{outSig} sigma) were found in loop delta:")
        for outlier in outliers:
            print(f"\tInd: {outlier} | Delta: {timedeltas[outlier]} us")

        print('Excluding the outliers, our stats improve to:')
        inForm = np.delete(timedeltas, outliers)
        print(f'Samples: {len(inForm)+1} | STD: {np.std(inForm):0.3f} us | Mean: {np.mean(inForm):0.3f} us')

    ######## Plotted statistics
    if isPlot:
        plt.ion()
        
        # A 0-referenced time vector for X axis of plots (in seconds)
        timeXAx = (timestamps-timestamps[0]) * NS2SEC

        plt.figure()
        plt.plot(np.arange(len(timedeltas)), timedeltas, 'o', markersize=2)
        plt.title(f'Time Deltas\nSamples: {len(timestamps)} | STD: {STD:0.3f} us | Mean: {Mean:0.3f} us')
        plt.xlabel('Loop Iteration')
        plt.ylabel('Loop Time [us]')

        plt.figure()
        plt.hist(timedeltas, bins=10)
        plt.xlabel('Loop Time [us]')
        plt.ylabel('Occurrences')

File: hstt/scripts/camerad-perf/cpp/TimingJitter_runner.cpp
#include <zmq.hpp>
#include <iostream>
#include <chrono>       // computing loop time (and sleep)
#include <fstream>      // saving csv file
#include <thread>

// Conversion Constants
int NS2US = 1e-3;       // nanosecond to microsecond conversion
int NS2SEC = 1e-9;      // nanosecond to second conversion

// User Inputs
int Nsamp = 10000;      // Size of preallocated timestamps array (max number of samples)
bool isPrint = false;   // Flag to print receive messages to terminal

int main(int argc, char** argv) {

    // Get filename for csv output, or note that we should just run local statistics
    bool isSaveCSV;
    std::string filename;
    if ( argc > 1 ) 
    {    
        isSaveCSV = true;
        std::cout << "Filename provided. Will save timestamps to: ";
        filename = argv[1];
        std::cout << filename << std::endl;
    } else {
        isSaveCSV = false;
        std::cout << "No Filename Provided... Will only print statistics" << std::endl;
    }

    // Preallocate output array
    long results[Nsamp] = {};

    // Create a ZeroMQ context
    zmq::context_t context(1);

    // Create an XSUB socket
    zmq::socket_t socket(context, zmq::socket_type::xsub);

    // Connect the socket to xpub
    socket.connect("tcp://localhost:5555");

    // Subscribe to all topics
    zmq::message_t subscribe_msg(1);
    memcpy(subscribe_msg.data(), "\x01", 1);
    socket.send(subscribe_msg, zmq::send_flags::none);

    std::cout << "Listening for messages on port 5555..." << std::endl;

    // Get starting timestamp
    auto t0 = std::chrono::high_resolution_clock::now();

    for (int sInd = 0; sInd < Nsamp; sInd++ ) {
        zmq::message_t message;
        socket.recv(message, zmq::recv_flags::none);
        //results[sInd] = std::chrono::duration<double>(std::chrono::high_resolution_clock::now()-t0).count();
        results[sInd] = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now()-t0).count();

        // std::cout << sInd << std::endl;

        std::this_thread::sleep_for(std::chrono::microseconds(1));

        if ( isPrint ) { 
            std::string received(static_cast<char*>(message.data()), message.size());
            std::cout << "data received: " << received << std::endl; 
        }
    }

    // Save CSV with timestamps
    if ( isSaveCSV ) {
        
        std::cout << "Saving to : " << filename << std::endl;

        std::ofstream csvFile(filename);
        if ( csvFile.is_open() ) {
            // Get starting precision for the output stream
            int prec = csvFile.precision();

            // Set precision high enough to record nanosecond time
            csvFile.precision(15);

            for (int sInd = 0; sInd < Nsamp; sInd++ ) {
                csvFile << results[sInd] << "\n";
                //std::cout << sInd << " : " << results[sInd] << std::endl;
            }
            csvFile.close();
        } else {
            std::cerr << "Unable to open CSV file for writing." << std::endl;
        }
    }

    return 0;
}
